crates/runtime/src/datafusion/extension/cache_invalidation.rs

```diff
@@ -0,0 +1,490 @@
+/// During physical planning, this node is converted into a [`CacheInvalidationExec`] execution plan,
+/// which performs cache invalidation after the write completes successfully.
+pub(crate) struct CacheInvalidationNode {
+    input: LogicalPlan,
+    table: TableReference,
+    caching: Weak<Caching>,
+}
+
+impl PartialOrd for CacheInvalidationNode {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        match self.input.partial_cmp(&other.input) {
+            Some(std::cmp::Ordering::Equal) => self.table.partial_cmp(&other.table),
+            non_eq => non_eq,
+        }
+    }
+}
+impl CacheInvalidationNode {
+    pub(crate) fn new(input: LogicalPlan, table: TableReference, caching: Weak<Caching>) -> Self {
+        Self {
+            input,
+            table,
+            caching,
+        }
+    }
+}
+
+impl Debug for CacheInvalidationNode {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        UserDefinedLogicalNodeCore::fmt_for_explain(self, f)
+    }
+}
+
+impl UserDefinedLogicalNodeCore for CacheInvalidationNode {
+    fn name(&self) -> &'static str {
+        "CacheInvalidationNode"
+    }
+
+    fn inputs(&self) -> Vec<&LogicalPlan> {
+        vec![&self.input]
+    }
+
+    fn schema(&self) -> &DFSchemaRef {
+        self.input.schema()
+    }
+
+    fn expressions(&self) -> Vec<Expr> {
+        Vec::new()
+    }
+
+    fn fmt_for_explain(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "CacheInvalidationNode: table={}", self.table)
+    }
+
+    fn with_exprs_and_inputs(&self, exprs: Vec<Expr>, inputs: Vec<LogicalPlan>) -> Result<Self> {
+        assert_eq!(inputs.len(), 1, "should have one input");
+        assert_eq!(exprs.len(), 0, "should have no expressions");
+        let Some(input) = inputs.into_iter().next() else {
+            unreachable!("should have one input");
+        };
+        Ok(Self {
+            input,
+            table: self.table.clone(),
+            caching: Weak::clone(&self.caching),
+        })
+    }
+
+    fn prevent_predicate_push_down_columns(&self) -> HashSet<String> {
+        // Allow filters for all columns to be pushed down
+        HashSet::new()
+    }
+
+    /// Returns the necessary input columns for this node required to compute
+    /// the columns in the output schema
+    ///
+    /// This is used for projection push-down when `DataFusion` has determined that
+    /// only a subset of the output columns of this node are needed by its parents.
+    /// This API is used to tell `DataFusion` which, if any, of the input columns are no longer
+    /// needed.
+    ///
+    /// Return `None`, the default, if this information can not be determined.
+    /// Returns `Some(_)` with the column indices for each child of this node that are
+    /// needed to compute `output_columns`
+    fn necessary_children_exprs(&self, _output_columns: &[usize]) -> Option<Vec<Vec<usize>>> {
+        // Default implementation
```

Comment:
Shouldn't we say that all of the input columns are required to compute the output columns?

Reply:
Updated, thank you. My thinking was that the node wraps write operations that returns hardcoded output schema (count) so does not require any columns to compute output schema